dex/encode/encode.go

```diff
@@ -0,0 +1,446 @@
+	rawCoins, err := ExtractPushes(coinsB)
+	if err != nil {
+		return nil, fmt.Errorf("error extracting coins: %v", err)
+	}
+	coins := make([]order.CoinID, 0, len(rawCoins))
+	for _, coin := range rawCoins {
+		coins = append(coins, coin)
+	}
+	sell := true
+	if bEqual(sellB, ByteFalse) {
+		sell = false
+	}
+	return &order.MarketOrder{
+		Coins:    coins,
+		Sell:     sell,
+		Quantity: IntCoder.Uint64(qtyB),
+		Address:  string(addrB),
+		Filled:   IntCoder.Uint64(filledB),
+	}, nil
+}
+
+// EncodeMatch encodes the UserMatch to bytes suitable for binary storage or
+// communications.
+func EncodeMatch(match *order.UserMatch) []byte {
+	return BuildyBytes{0}.
+		AddData(match.OrderID[:]).
+		AddData(match.MatchID[:]).
+		AddData(Uint64Bytes(match.Quantity)).
+		AddData(Uint64Bytes(match.Rate)).
+		AddData([]byte(match.Address)).
+		AddData(Uint64Bytes(match.Time)).
+		AddData([]byte{byte(match.Status)}).
+		AddData([]byte{byte(match.Side)})
+}
+
+// DecodeMatch decodes the versioned blob into a UserMatch.
+func DecodeMatch(b []byte) (match *order.UserMatch, err error) {
+	err = VersionedDecode("decodeMatch", b, map[byte]ByteDecoder{
+		0: func(pushes [][]byte) error {
+			match, err = matchDecoder_v0(pushes)
+			return err
+		},
+	})
+	return match, err
+}
+
+// matchDecoder_v0 decodes the version 0 payload into a *UserMatch.
+func matchDecoder_v0(pushes [][]byte) (*order.UserMatch, error) {
+	if len(pushes) != 8 {
+		return nil, fmt.Errorf("decodeMatch: expected 8 pushes, got %d", len(pushes))
+	}
+	oidB, midB := pushes[0], pushes[1]
+	if len(oidB) != order.OrderIDSize {
+		return nil, fmt.Errorf("decodeMatch: expected length %d order ID, got %d", order.OrderIDSize, len(oidB))
+	}
+	if len(midB) != order.MatchIDSize {
+		return nil, fmt.Errorf("decodeMatch: expected length %d match ID, got %d", order.MatchIDSize, len(midB))
+	}
+	var oid order.OrderID
+	copy(oid[:], oidB)
+	var mid order.MatchID
+	copy(mid[:], midB)
+
+	statusB, sideB := pushes[6], pushes[7]
+	if len(statusB) != 1 || len(sideB) != 1 {
+		return nil, fmt.Errorf("decodeMatch: status/side incorrect length %d/%d", len(statusB), len(sideB))
+	}
+	return &order.UserMatch{
+		OrderID:  oid,
+		MatchID:  mid,
+		Quantity: IntCoder.Uint64(pushes[2]),
+		Rate:     IntCoder.Uint64(pushes[3]),
+		Address:  string(pushes[4]),
+		Time:     IntCoder.Uint64(pushes[5]),
+		Status:   order.MatchStatus(statusB[0]),
+		Side:     order.MatchSide(sideB[0]),
+	}, nil
+}
+
+// Single-length byte slices used as a flag to indicate common order constants.
+var (
+	orderTypeLimit    = []byte{'l'}
+	orderTypeMarket   = []byte{'m'}
+	orderTypeCancel   = []byte{'c'}
+	orderTifImmediate = []byte{'i'}
+	orderTifStanding  = []byte{'s'}
+)
+
+// EncodeOrder encodes the order to bytes sutiable for wire communications or
+// database storage. EncodeOrder accepts any type of Order.
+func EncodeOrder(ord order.Order) []byte {
```

Comment:
These order serialization methods could be moved to the order package. In the long term, I'm hoping to use these types of serializations both for DEX signature messages and as the basis for a post-JSON messaging protocol.